astropy/time/tests/test_precision.py

```diff
@@ -142,69 +293,368 @@ def test_leap_seconds_rounded_correctly():
+def test_day_frac_always_less_than_half(jds):
+    jd1, jd2 = jds
+    t_jd1, t_jd2 = day_frac(jd1, jd2)
+    assert np.all(t_jd1 % 1 == 0)
+    assert np.all(abs(t_jd2) <= 0.5)
+    assert np.all((abs(t_jd2) < 0.5) | (t_jd1 % 2 == 0))
+
+
+@given(integers(-10**8, 10**8), sampled_from([-0.5, 0.5]))
+def test_day_frac_round_to_even(jd1, jd2):
+    t_jd1, t_jd2 = day_frac(jd1, jd2)
+    assert (abs(t_jd2) == 0.5) and (t_jd1 % 2 == 0)
+
+
+@given(scale=sampled_from(STANDARD_TIME_SCALES), jds=unreasonable_jd())
+@example(scale="tai", jds=(0.0, 0.0))
+@example(scale="tai", jds=(0.0, -31738.500000000346))
+def test_resolution_never_decreases(scale, jds):
+    jd1, jd2 = jds
+    assume(not scale == 'utc' or 2440000 < jd1 + jd2 < 2460000)
+    t = Time(jd1, jd2, format="jd", scale=scale)
+    with quiet_erfa():
+        assert t != t + dt_tiny
+
+
+@given(reasonable_jd())
+def test_resolution_never_decreases_utc(jds):
+    """UTC is very unhappy with unreasonable times"""
+    jd1, jd2 = jds
+    t = Time(jd1, jd2, format="jd", scale="utc")
+    with quiet_erfa():
+        assert t != t + dt_tiny
+
+
+@given(scale1=sampled_from(STANDARD_TIME_SCALES),
+       scale2=sampled_from(STANDARD_TIME_SCALES),
+       jds=unreasonable_jd())
+@example(scale1='tcg', scale2='ut1', jds=(2445149.5, 0.47187700984387526))
+@example(scale1='tai', scale2='tcb', jds=(2441316.5, 0.0))
+@example(scale1='tai', scale2='tcb', jds=(0.0, 0.0))
+def test_conversion_preserves_jd1_jd2_invariant(iers_b, scale1, scale2, jds):
+    jd1, jd2 = jds
     t = Time(jd1, jd2, scale=scale1, format="jd")
-    with pytest.warns(
-            ErfaWarning,
-            match=r'ERFA function "taiutc" yielded 1 of "dubious year'):
-        t2 = getattr(t, scale2)
+    try:
+        with quiet_erfa():
+            t2 = getattr(t, scale2)
+    except iers.IERSRangeError:  # UT1 conversion needs IERS data
+        assume(False)
+    except ErfaError:
+        assume(False)
     assert t2.jd1 % 1 == 0
     assert abs(t2.jd2) <= 0.5
     assert abs(t2.jd2) < 0.5 or t2.jd1 % 2 == 0
 
 
-def test_conversion_preserves_jd1_jd2_invariant_2():
-    """Conversion can leave abs(jd2)>0.5"""
-    scale1 = 'tai'
-    scale2 = 'tcb'
-    jd1, jd2 = (2441316.5, 0.0)
+@pytest.mark.xfail(reason='better xfail than an unknown bug', strict=True)
+@given(scale1=sampled_from(STANDARD_TIME_SCALES),
+       scale2=sampled_from(STANDARD_TIME_SCALES),
+       jds=unreasonable_jd())
+def test_conversion_never_loses_precision(iers_b, scale1, scale2, jds):
+    jd1, jd2 = jds
     t = Time(jd1, jd2, scale=scale1, format="jd")
-    t2 = getattr(t, scale2)
-    assert t2.jd1 % 1 == 0
-    assert abs(t2.jd2) <= 0.5
-    assert abs(t2.jd2) < 0.5 or t2.jd1 % 2 == 0
+    # If the rates ever differ, might lose one ULP
+    with quiet_erfa():
+        t2 = t + 2*dt_tiny
+        try:
+            assert getattr(t, scale2) < getattr(t2, scale2)
+        except iers.IERSRangeError:  # UT1 conversion needs IERS data
```

Comment:
I guess we'd need a separate `except ERFAError` for `utc` here to be able to remove the `xfail`. This happens for years so early that JD cannot be computed:  `jd1+jd2 < -31738.5`.

Reply:
```python
Falsifying example: test_conversion_never_loses_precision(
    iers_b='<using IERS-B orientation table>',
    scale1='tai',
    scale2='tcb',
    jds=(-31738.500000000346, 0.0),
)
```

Not just UTC either :confused:

Reply:
Ah, yes, to convert to `tdb` or `tcb`, one needs a `delta_tdb_tt` which cannot be calculated for those early times, as it needs UTC.  See figure in https://docs.astropy.org/en/latest/time/index.html#time-scale

I guess we should step back and think about the purpose of this test. Presumably just to check accuracy for allowed transformations, which for the broad range covered here is only between TAI, TT, and TCG and between TDB and TCB.

Reply:
Ah, that's done it - thanks @mhvk!